### ISSUE TITLE
Weird crash when giving an existing_clustering to hierarchical_clustering

### ISSUE THREAD

**Issue by Pachidermus** [fsavje/Rscclust#12]
Hello there, 

I'm using the latest version of `scclust` (0.2.2) and `distances` (0.1.8) to construct many size-constrained clusters.
As I want to get the smallest possible clusters while respecting the minimum size constraint, I first use `sc_clustering` then `hierarchical_clustering` which works flawlessly for most of the groups.

However, I noticed that, on certain groups, calling `hierarchical_clustering` with an `existing_clustering` attribute crashes my R session. 

You may find here: [issue_data.csv](https://github.com/fsavje/scclust-R/files/8573714/issue_data.csv), data that produces a crash on the following code snippet. 

```
library(distances)
library(scclust)

X <- read.csv2("issue_data.csv") %>% as.matrix()

X_distances <- distances(X)
clustering <- sc_clustering(X_distances, size_constraint=10, seed_method="inwards_updating")

# crashes happen with the following line
h_clustering <- hierarchical_clustering(X_distances, size_constraint=10, existing_clustering=clustering)
```

The problem doesn't seem to be related to memory issues since the snippet above works on way bigger groups.
Moreover, the final line works when no `existing_clustering` is given. 
I tried to run `hierarchical_clustering` on each cluster produced by `clustering` independently and met no issue. 

Thank you for your support.

**Comment by fsavje** [comment]
Hi @Pachidermus 

Thanks for your bug report. I haven't been able to track down exactly where the problem is, but it is related to the fact that you have very many features in your clustering problem.

I would **strongly** recommend you to do some type of dimensionality reduction in the feature space before clustering here. There is no clustering algorithm that can handle 250+ features when there is ~500 observations unless the features are highly correlated (in which case dimensionality reduction is straightforward any way). The problem is that the data points will be close to equidistant when there are many features, so the distances become meaningless. As a rule of thumb, with ~500 observations, about 10 features is the upper limit. This is true for all clustering approaches. (There are some clustering approaches that have built in dimensionality reduction, but most doesn't.)

Now, if you still want to cluster on this many features and don't want to wait for me to track down the problem, you can use the following work around code:

```{r}
library(distances)
library(scclust)

# Alternative hierarchical clustering function
alt_hierarchical_clustering <- function(df, size_constraint, existing_clustering) {
  make_h_cl <- function(label, df) {
    paste0(label, ".", hierarchical_clustering(distances(as.matrix(df)), size_constraint))
  }
  out <- character(length(existing_clustering))
  split(out, existing_clustering) <-
    mapply(make_h_cl, sort(unique(existing_clustering)), split(X, existing_clustering))
  out
}

# Read data as data frame
X <- read.csv2("issue_data.csv")

# Make distances object and call `sc_clustering()`
X_distances <- distances(as.matrix(X))
cl <- sc_clustering(X_distances, size_constraint = 10, seed_method = "inwards_updating")

# Call alternative hierarchical clustering function (note that data frame is used)
h_cl <- alt_hierarchical_clustering(X, size_constraint = 10, existing_clustering = cl)

# Compare clusterings
table(cl, h_cl)

# Transform clustering to scclust object if needed
h_cl <- scclust(h_cl)
```

**Comment by Pachidermus** [comment]
@fsavje Thank you for your quick answer !

I had already found a less elegant work-around  by the time I decided to write down this issue.
Thank you for sharing this code, that might also help future users :) 

I can't help but agree with you on that matter: in many problems, clustering with this many dimensions doesn't make that much sense. I'm actually working on a weird scenario on which I would like to avoid using more dimensionality reduction techniques like PCA. 
As a matter of fact, I probably didn't give enough precision on what I was trying to do and where my data came from. 
My observations (which were long time series of +15k timesteps at the beginning) already went through a dimensionality reduction process through the usage of convolutional autoencoders.

Long story short, I'm trying out different sizes of reduction (292 being the largest as you probably expect) and different techniques in order to build a space in which clustering would perform okay. In the extract I gave you, there are only about 500 observations but on other groups, there are more than 10k.

Thank you again for your time and support!
Keep up the good work 👍
